cognite_synthetic_tags/data_stores.py

```diff
@@ -0,0 +1,287 @@
+from typing import Any, List, Literal, Optional, Set
+
+import numpy as np
+import pandas as pd
+from cognite.client import CogniteClient
+from cognite.client.utils._time import granularity_to_ms, timestamp_to_ms
+
+from cognite_synthetic_tags.types import (
+    CogniteTimeT,
+    RetrievalFuncT,
+    TagValueStoreResultT,
+)
+
+__all__ = [
+    "retrieve_datapoints_df",
+    "series",
+    "point",
+    "series_at_time",
+    "point_at_time",
+]
+
+
+def retrieve_datapoints_df(
+    c: CogniteClient,
+    id: Optional[List[int]] = None,
+    external_id: Optional[List[str]] = None,
+    ignore_unknown_ids: bool = True,
+    fill_with=np.nan,
+    **kwargs,
+) -> pd.DataFrame:
+    """
+    Wrapper around `CogniteClient.datapoints.retrieve` which fills in any
+    missing columns (i.e. values in external_ids).
+
+    Calling, `CogniteClient.datapoints.retrieve(..., ignore_unknown_ids=True)`
+    results in missing columns, while `ignore_unknown_ids="fill"` or
+    `"fill,dropna"` raise exceptions.
+    """
+    res = c.datapoints.retrieve(
+        id=id,
+        external_id=external_id,
+        ignore_unknown_ids=ignore_unknown_ids,
+        **kwargs,
+    )
+    df = res.to_pandas()
+
+    # Compile a list of expected columns:
+    tags = external_id or id or []
+    aggregates = kwargs.get("aggregates")
+    columns = (
+        tags
+        if aggregates is None
+        else [f"{tag}|{aggregate}" for tag in tags for aggregate in aggregates]
+    )
+
+    # Add any missing columns:
+    df.loc[:, [c for c in columns if c not in df.columns]] = fill_with
+
+    # Reorder columns:
+    df = df[columns]
+
+    return df.copy()
```

Comment:
Could be wrong (I'm rather new to Pandas), but I think `df[columns]` will return a "view". So later, after the function returns, things like `df["foo"] = bar` might emit warnings about settings values on a view rather than on a real dataframe. Again, I could easily be wrong...

Reply:
Ok, can't find a reason for the `.copy()`, removed it.